contrib/TestHarness2/test_harness/results.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+import argparse
+import io
+import json
+import re
+import sys
+import test_harness.fdb
+
+from typing import List, Tuple, OrderedDict
+from test_harness.summarize import SummaryTree, Coverage
+from test_harness.config import config
+from xml.sax.saxutils import quoteattr
+
+
+class GlobalStatistics:
+    def __init__(self):
+        self.total_probes_hit: int = 0
+        self.total_cpu_time: int = 0
+        self.total_test_runs: int = 0
+        self.total_missed_probes: int = 0
+
+
+class EnsembleResults:
+    def __init__(self, cluster_file: str | None, ensemble_id: str):
+        self.global_statistics = GlobalStatistics()
+        self.fdb_path = ('joshua', 'ensembles', 'results', 'application', ensemble_id)
+        self.coverage_path = self.fdb_path + ('coverage',)
+        self.statistics = test_harness.fdb.Statistics(cluster_file, self.fdb_path)
+        coverage_dict: OrderedDict[Coverage, int] = test_harness.fdb.read_coverage(cluster_file, self.coverage_path)
+        self.coverage: List[Tuple[Coverage, int]] = []
+        self.min_coverage_hit: int | None = None
+        self.ratio = self.global_statistics.total_test_runs / config.hit_per_runs_ratio
+        for cov, count in coverage_dict.items():
+            if re.search(config.cov_include_files, cov.file) is None:
+                continue
+            if re.search(config.cov_exclude_files, cov.file) is not None:
+                continue
+            self.global_statistics.total_probes_hit += count
+            self.coverage.append((cov, count))
+            if count <= self.ratio:
+                self.global_statistics.total_missed_probes += 1
+            if self.min_coverage_hit is None or self.min_coverage_hit > count:
+                self.min_coverage_hit = count
+        self.coverage.sort(key=lambda x: (x[1], x[0].file, x[0].line))
+        self.stats: List[Tuple[str, int, int]] = []
+        for k, v in self.statistics.stats.items():
+            self.global_statistics.total_test_runs += v.run_count
+            self.global_statistics.total_cpu_time += v.runtime
+            self.stats.append((k, v.runtime, v.run_count))
+        self.stats.sort(key=lambda x: x[1], reverse=True)
+        if self.min_coverage_hit is not None:
+            self.coverage_ok = self.min_coverage_hit > self.ratio
```

Comment:
Yes and for now I wanted to leave it this restrictive (and fail more often that it probably should). Mostly because I am planning to work on coverage misses next. For now this will always fail anyways (as it seems a lot of probes are simply not reachable).

But you're right we should redefine this later (including the once per 20k is probably too rigit in the first place). But basically I need to collect some more data and experience before writing this code.